### PR TITLE
OH2-409 | Add lock field to medical ward

### PR DESCRIPTION
**Pull request by SteveGT96**
See [OH2-409](https://openhospital.atlassian.net/browse/OH2-409)

[OH2-409]: https://openhospital.atlassian.net/browse/OH2-409?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sql/step_a112_add_medical_ward_lock_column.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+--
```
Comment (dbmalkovsky) [inline sql/step_a112_add_medical_ward_lock_column.sql]: does the demo data file need to be updated too?

File: src/main/java/org/isf/medicalstockward/model/MedicalWard.java [hunk 2]
```diff
@@ -45,45 +46,51 @@
...
+	@Column(name = "MDSRWRD_OUT_QTI")
 	private float out_quantity;
-	
+
+	/**
+	 * Lock control
+	 */
+	@Version
+	@Column(name = "MDSRWRD_LOCK")
+	private Integer lock;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalstockward/model/MedicalWard.java]: why is this an `Integer`?; isn't `int` OK and better as it is a primitive and not an object.

Comment (SteveGT96) [reply src/main/java/org/isf/medicalstockward/model/MedicalWard.java]: You're right. It was even making tests failing.

File: sql/step_a112_add_medical_ward_lock_column.sql (hunk 1 above)
Comment (SteveGT96) [reply sql/step_a112_add_medical_ward_lock_column.sql]: Yes !

**Comment by SilverD3** [comment]
@SteveGT96 is it possible to add some tests to ensure that lock is effectively updated when we update a 'ward medical'?

File: src/test/java/org/isf/medicalstock/Tests.java [hunk 3]
```diff
@@ -152,12 +153,6 @@ static void setUpClass() {
 		testSupplier = new TestSupplier();
 		testMedicalStock = new TestMedicalStock();
 	}
-
-	@BeforeEach
-	void setUp() throws OHException {
-		cleanH2InMemoryDb();
-	}
-
 	@AfterAll
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/medicalstock/Tests.java:157]: Why did you remove the blank line before this method?

File: src/test/java/org/isf/medicalstock/Tests.java [hunk 4]
```diff
@@ -169,7 +164,10 @@ static void tearDownClass() {
 		testSupplier = null;
 		testMedicalStock = null;
 	}
-
+	@BeforeEach
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/medicalstock/Tests.java:169]: and here another blank line was removed

File: src/test/java/org/isf/medicalstock/Tests.java [hunk 5]
```diff
@@ -169,7 +164,10 @@ static void tearDownClass() {
...
+	void setUp() throws OHException {
+		cleanH2InMemoryDb();
+	}
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/medicalstock/Tests.java:172]: And a blank line after this method is gone too.

File: src/test/java/org/isf/medicalstock/Tests.java [hunk 6]
```diff
@@ -1529,6 +1527,20 @@ void testIoUpdateMedicalStockTableEmptyTableWithNegativeQuantity() throws Except
...
+	@Test
+	void testMedicalWardLockGetterSetter() throws Exception {
+		int code = setupTestMovement(false);
+		Movement movement = movementIoOperationRepository.findById(code).orElse(null);
+		assertThat(movement).isNotNull();
+		MovementType movementType = movement.getType();
+		movementType.setType("-");
+		MedicalWard medicalWard = new MedicalWard(movement.getWard(), movement.getMedical(), 0, 0, movement.getLot());
+
+		assertThat(medicalWard.getLock()).isEqualTo(0);
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/medicalstock/Tests.java]:
```suggestion
		assertThat(medicalWard.getLock()).isZero();
```

**Review by dbmalkovsky — COMMENTED** [review]
left a few minor comments

File: src/test/java/org/isf/medicalstock/Tests.java (hunk 3 above)
Comment (SteveGT96) [reply src/test/java/org/isf/medicalstock/Tests.java:157]: That may be due to the ide reformat. Just fixed it.

File: sql/step_04_all_following_steps.sql [hunk 7]
```diff
@@ -105,3 +105,4 @@ source step_a108_audit_dicom_type_and_dicom_data.sql;
 source step_a109_update_user_settings_table_constraints_and_add_usergroups_permissions.sql;
 source step_a110_update_operations_table_change_ope_for_to_enum.sql;
 source step_a111_add_missing_lock_columns.sql;
+source step_a112_add_medical_ward_lock_column.sql;
```
Comment (mwithi) [inline sql/step_04_all_following_steps.sql]: Sorry, noticed now, can you use the already existing `step_a111` for this purpose?
It is conflicting because in CORE we already have this situation:

```
...
source step_a109_update_user_settings_table_constraints_and_add_usergroups_permissions.sql;
source step_a110_update_operations_table_change_ope_for_to_enum.sql;
source step_a111_add_missing_lock_columns.sql;
source step_a112_users_and_groups_soft_deletion.sql;

```

Comment (SteveGT96) [reply sql/step_04_all_following_steps.sql]: OK !

File: src/test/java/org/isf/medicalstock/Tests.java [hunk 8]
```diff
@@ -1106,7 +1107,7 @@ void testMgrValidateLotWithCostZero() {
 			Movement movement = movementIoOperationRepository.findById(code).orElse(null);
 			assertThat(movement).isNotNull();
 			Lot lot = movement.getLot();
-			lot.setCost(new BigDecimal(0.0));
+			lot.setCost(new BigDecimal("0.0"));
```
Comment (mwithi) [inline src/test/java/org/isf/medicalstock/Tests.java]:
```suggestion
			lot.setCost(BigDecimal.ZERO);
```

File: src/main/java/org/isf/medicalstockward/model/MedicalWard.java [hunk 9]
```diff
@@ -45,45 +46,51 @@
...
-	
+
+	/**
+	 * Lock control
+	 */
+	@Version
+	@Column(name = "MDSRWRD_LOCK", columnDefinition = "INT(11) NOT NULL DEFAULT 0")
+	private int lock;
 	@Transient
 	private Double qty = 0.0;
```
Comment (mwithi) [inline src/main/java/org/isf/medicalstockward/model/MedicalWard.java]:
```suggestion
	@Version
	@Column(name = "MDSRWRD_LOCK", columnDefinition = "INT(11) NOT NULL DEFAULT 0")
	private int lock;
	
	@Transient
	private Double qty = 0.0;
```

File: src/main/java/org/isf/medicalstockward/model/MedicalWard.java [hunk 10]
```diff
@@ -92,93 +99,97 @@ public MedicalWard(Medical med, double qanty, Lot lot) {
...
 	public void setId(Ward ward, Medical medical, Lot lot) {
-		this.id = new MedicalWardId(ward, medical, lot); 
+		this.id = new MedicalWardId(ward, medical, lot);
 	}
-	
+
 	public Lot getLot() {
 		return id.getLot();
 	}
-
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalstockward/model/MedicalWard.java:103]: I would leave a line after the end of the method.

File: src/main/java/org/isf/medicalstockward/model/MedicalWard.java [hunk 11]
```diff
@@ -92,93 +99,97 @@ public MedicalWard(Medical med, double qanty, Lot lot) {
...
 	public Lot getLot() {
 		return id.getLot();
 	}
-	
-	public void setMedical(Medical medical) {
-		this.id.setMedical(medical);
+	public void setLot(Lot lot) {
+		id.setLot(lot);
 	}
-
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalstockward/model/MedicalWard.java:107]: I would leave a line after the end of the method.

File: src/main/java/org/isf/medicalstockward/model/MedicalWard.java [hunk 12]
```diff
@@ -92,93 +99,97 @@ public MedicalWard(Medical med, double qanty, Lot lot) {
...
-	
+
 	public void setWard(Ward ward) {
 		this.id.setWard(ward);
 	}
 
 	public Medical getMedical() {
 		return this.id.getMedical();
 	}
-
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/medicalstockward/model/MedicalWard.java:137]: I would leave a line after the end of the method.

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!
